### ISSUE TITLE
one logger per engine declared in settings.yml ?

### ISSUE THREAD

**Issue by dalf** [searxng/searxng#98]
Should a logger be created automatically for each engine in `searx.engines.__init__` ?

For example:
https://github.com/searxng/searxng/blob/8283ce3d79b67bad0820dad7e02ab8140ec44898/searx/engines/xpath.py#L28

Is it interesting to have one logger per engine declared in settings.yml ?

Downside: it adds one more magical variable, that IDE / editors won't know. Solution: use a factory `logger = searx.engines.get_logger()`.

**Comment by return42** [comment]
> Should a logger be created automatically for each engine in searx.engines.init ?

could be convenient  .. especial with the generic engines (XPath).

> Is it interesting to have one logger per engine declared in settings.yml ?

Why in the settings? .. why not just attaching `logger` function to the namespace of the engine, when calling `searx.engines.load_engine(..)` ?

> Downside: it adds one more magical variable, 

engines are *special* (the namespaces of engines & plugins are *special*) .. I think it is acceptable.

> Solution: use a factory `logger = searx.engines.get_logger()`.

Hm .. where should the engine name come from .. and what is the difference to:

https://github.com/searxng/searxng/blob/8283ce3d79b67bad0820dad7e02ab8140ec44898/searx/engines/xpath.py#L28

**Comment by dalf** [comment]
>  why not just attaching logger function to the namespace of the engine, when calling searx.engines.load_engine(..) ?

Yes that's exactly what I was trying to describe:

in searx/engines/__init__.py : 
```python
def load_engine(engine_data):
    ...
    engine.logger = logger.getChild('searxng.engines.' + engine_name)
```

`logger` must be added to https://github.com/searxng/searxng/blob/7c76cef746a90542f7db7a9ca876e763d0aad1a3/manage#L36


> Hm .. where should the engine name come from

```python
def load_engine(engine_data):
    ...
    engine.get_logger = lambda: logger.getChild('searxng.engines.' + engine_name)
```

then xpath.py
```
logger = get_logger()
```

since `get_logger()` will be different for each engine, it will work the same way the first solution ( `engine.logger = ...`).

Here the only benefit is when the code is edited (depending of the editor).

-->

The first solution is more in the "spirit" of the engine definitions.

**Comment by dalf** [comment]
About :
```python
logger.debug("query_url --> %s", query_url)
```

httpx already log the HTTP request:
```
DEBUG:httpx._client:HTTP Request: GET https://bitbucket.org/repo/all/1?name=test "HTTP/2 200 OK"
```

If there is one logger per engine, what can be done:
```
DEBUG: searxng.engine.bitbucket: GET https://bitbucket.org/repo/all/1?name=test "HTTP/2 200 OK"
```
(and hide the `httpx._client` log message).

So there is no need to add `logger.debug("query_url --> %s", query_url)` in each engine.

**Comment by return42** [comment]
> httpx already log the HTTP request:

Ah, this is a good point .. since httpx we do no longer need my "query_url -->" logs .. good to know ..  I will take this in mind.

> and hide the httpx._client log message

I don't have a clue right now how to do this .. if you have a solution for .. send a PR and lets discuss further in context of the code changes.

**Comment by dalf** [comment]
> I don't have a clue right now how to do this .. if you have a solution for .. send a PR and lets discuss further in context of the code changes.

I usual use the httpx._client log. If this is enough we can just remove the `logger.debug("query_url --> %s", query_url)`.
For the log per engine solution:
https://github.com/searxng/searxng/blob/b48b4c93d586f52418295a15a1c93934f007c41f/searx/network/client.py#L234-L235

**Comment by dalf** [comment]
Closed by #296
